### ISSUE TITLE
Connection ID leaks when some error occurs

### ISSUE THREAD

**Issue by ChenHuajun** [citusdata/citus#2526]
After the following error occurs, citus will miss the release of Connection IDs. When the 2048 elements of ClientConnectionArray are all full, the "could not allocate connection in connection pool" error occurs.

```
ERROR:  cannot establish a new connection for placement 11440, since DML has been executed on a connection that is in use
```

## environment
- PostgreSQL 10.2
- citus 7.2.1 and citus 7.4.1

## How to reproduct

1. Make "cannot establish a new connection for placement" error happen
```
lobadbt=# begin;
BEGIN
lobadbt=# update tb1 set id=id where id=1;
UPDATE 0
lobadbt=# update tb1 set id=id where id=3; -- The two records with id values 1 and 3 are located on 2 different shards of the same worker.
UPDATE 0
lobadbt=#  select * from tb1 where id in(1,3);
ERROR:  cannot establish a new connection for placement 11440, since DML has been executed on a connection that is in use
```

2. Check the value of the ClientConnectionArray array and find that an element remains uncleared (normally, the elements in the ClientConnectionArray will be cleared after the SQL ends)
```
[postgres@lobapstdb158 ~]$ gdb -ex 'p ClientConnectionArray' --batch -p 991
[Thread debugging using libthread_db enabled]
0x000000369e2e8f13 in __epoll_wait_nocancel () from /lib64/libc.so.6
$1 = {0x2b903e8, 0x0 <repeats 2047 times>}
```

3. Repeat the above steps, the remaining elements in the ClientConnectionArray will increase
```
[postgres@lobapstdb158 ~]$ gdb -ex 'p ClientConnectionArray' --batch -p 991
[Thread debugging using libthread_db enabled]
0x000000369e2e8f13 in __epoll_wait_nocancel () from /lib64/libc.so.6
$1 = {0x2b903e8, 0x2b903e8, 0x0 <repeats 2046 times>}
```

**Comment by marcocitus** [comment]
Thanks for the detailed analysis. We had an open issue, but closing it in favour of this one.

**Comment by marcocitus** [comment]
Fixed by #2798
